### PR TITLE
fix(dpd.js): parse script src for root url, apposed to window.location

### LINKED ISSUE #83: Defining an API Root
It would be pretty useful if you could define a different point for which API requests would be responded to. For example:

before:
GET /contacts
GET /users

after having defined a different API root: (/api/)
GET /api/contacts
GET /api/users

### PR DESCRIPTION
**Pull request by sscovil**
enable users to host deployd on a subdomain (e.g. api.mydomain.com) and still use dpd.js

might resolve #83 & #149, see https://groups.google.com/forum/#!topic/deployd-users/CT84hulBO40

### CODE REVIEW

**Comment by sscovil** [comment]
I know there have been a lot of discussions and a couple of pull requests regarding adding customizable endpoint paths, but this pull is a lightweight fix that simply ensures `dpd.js` can be used in conjunction with a proxy pass in Nginx or Apache.

See the following discussion in the Deployd user group for details: https://groups.google.com/forum/#!topic/deployd-users/CT84hulBO40

**Comment by sscovil** [comment]
After testing on IE, I found that `document.currentScript` is not supported and had to instead use `document.getElementsByTagName('src')`, looping through the results until the `/dpd.js` src is found.

As a result, I decided to leave the original `root` variable declaration in place as a failsafe. If some crazy how my technique fails, root will still have a value. I'm not crazy about that sort of defensive coding, but I didn't want to run the risk leaving root undefined...thoughts?

**Comment by ericfong** [comment]
please take a look 
#490 and
#437

**Comment by sscovil** [comment]
@ericfong I saw those and, while I see value in being able to configure the root directory, I believe the default use of `window.location.protocol`, `window.location.hostname` and `window.location.port` is flawed. Wherever I import `dpd.js` from should, in my opinion, be the default root for API calls.

I think this change could be incorporated into #490, but could be used on its own if that pull won't make it into the next release. What do you think?

**Comment by ericfong** [comment]
I think dpd.js may not called ```dpd.js```

Like me,
My build script download latest dpd.js from server
And compress it with all other codes to build a 'app.bundle.js'.
And I distribute that within my cordova app.

Why?
that is my quick fix to prevent dpd become undefined while mobile is offline.

**Comment by sscovil** [comment]
@ericfong I think yours is a good case for configurable root variables, but I still think the existing code that uses `window.location` is broken and should be fixed sooner than later.

Using the `dpd.js` script src, however it is parsed, is more accurate than the current method which does not allow for proxy-ing an instance of Deployd. If the configurable solution presented in #490 doesn't make it into the next release, this (or some variation of this) should IMHO.

**Comment by sscovil** [comment]
@ericfong My latest commit uses a series of methods to determine the script src and is flexible enough to handle your case of having a different file name.

* First, it attempts to use `document.currentScript` which is the most efficient method, but not supported in IE or older browsers.
* Next, it looks for an element with `id="dpd.js"` that also has a `src` attribute. This is how you can indicate the correct script in your case.
* Finally, it loops through all script tags until it finds the one ending in `dpd.js`. This is least efficient, but will work in any browser.

Once the script element is found, any .js filename is removed from the `src` value (along with the trailing slash) and the remaining string is assigned to `var root`.

File: clib/dpd.js [hunk 1]
```diff
@@ -2,9 +2,31 @@
 
   if (!window._dpd) window._dpd = {};
 
-  var root = window.location.protocol + '//' + window.location.hostname;
-  if (window.location.port !== '') {
-    root += ':' + window.location.port;
+  var root = dpdScriptElement().getAttribute('src')
+      .replace(/(.*)(\/.*\.js)$/, '$1')
+      .replace(/^[^\/]*.js$/, '');
+
```
Comment (ericfong) [inline clib/dpd.js]: Is it possible that root still not yet set in here? 

No currentScript, no id=dpd.js and script src is changed?

if yes, let's add a checking and fail back to the old code?

**Comment by ericfong** [comment]
BTW, do you think use a function to setup the root Url is better ?

Or use a auto parse way?

We can raise that to 0.8. if more people think that is necessary

File: clib/dpd.js (hunk 1 above)
Comment (sscovil) [inline clib/dpd.js]: Good catch. Actually, I'm thinking a polyfill for document.currentScript might be the best way to go here...this looks promising: https://github.com/JamesMGreene/document.currentScript

Comment (ericfong) [inline clib/dpd.js]: We better don't add dependencies at this stage.

File: clib/dpd.js [hunk 2]
```diff
@@ -2,9 +2,31 @@
...
+  function dpdScriptElement() {
+    // Only works in modern non-IE browsers
+    var element = document.currentScript;
+
+    if (!element) {
+      // Allows script to have a name other than dpd.js by assigning id="dpd.js" to script tag
+      element = document.getElementById('dpd.js');
```
Comment (ericfong) [inline clib/dpd.js]: This is a bit strange to me. 
As most people won't setup like this id="dpd.js". 

Please remove document.getElementById('dpd.js') to save some processing time.

File: clib/dpd.js [hunk 3]
```diff
@@ -2,9 +2,31 @@
...
+    }
+
+    if (!element || !element.hasAttribute('src')) {
+      // As a last resort, loop through all script tags and find the src attribute ending in "dpd.js"
+      var scriptTags = document.getElementsByTagName('script');
+      for (var i = 0; i < scriptTags.length; i++) {
+        if (scriptTags[i].hasAttribute('src') && scriptTags[i].getAttribute('src').match(/.*dpd\.js$/)) {
```
Comment (ericfong) [inline clib/dpd.js]: use document.querySelectorAll('script[src$="dpd.js"]') ?

**Comment by andreialecu** [comment]
Another possible idea could be to use an attribute like `data-dpd-host="auto"` or with a URL specified instead of auto, used on the script tag. Auto would mean it is implied from the src.

**Comment by ericfong** [comment]
Actually, I prefer rebase #490  which originally coded by @iranreyes and @andreialecu gist.

That PR should solve this problem by calling js function instead of adding a special to the html

**Comment by drewdotpro** [comment]
Please, for the love of God pull this pull request.
11/10

**Comment by andreialecu** [comment]
I think we'll opt for #490 instead, which incorporates these changes and is more powerful.

**Comment by drewdotpro** [comment]
OK good luck. We moved over to use sailsjs.org as it's (sorry) way better as an API building tool.

**Comment by andreialecu** [comment]
This has been fixed by merging PR #490 which includes the functionality of this PR. Closing this.

**Pull request closed without merging**